### PR TITLE
Implement ErrorReport API

### PR DESCRIPTION
**Pull request by kashike**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1215) | [**Common**](https://github.com/SpongePowered/SpongeCommon/pull/689)

### CODE REVIEW

File: src/main/java/org/spongepowered/common/SpongeBootstrap.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
  */
 package org.spongepowered.common;
 
+
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/SpongeBootstrap.java]: Unnecessary addition.

File: src/main/java/org/spongepowered/common/mixin/core/crash/MixinCrashReport.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import java.util.concurrent.CompletableFuture;
+
+@Mixin(CrashReport.class)
+public abstract class MixinCrashReport implements ErrorReport, IMixinCrashReport {
+
+    @Shadow private String description;
+
+    @Shadow private Throwable cause;
+
+    @Shadow public abstract CrashReportCategory makeCategory(String name);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/crash/MixinCrashReport.java]: organize these into fields then methods please.

File: src/main/java/org/spongepowered/common/service/error/GistPostingCallable.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import java.net.URLConnection;
+import java.nio.charset.Charset;
+import java.util.concurrent.Callable;
+
+public class GistPostingCallable implements Callable<URL> {
+
+    private static final URL GIST_POST_URL;
+    static {
+        try {
+            GIST_POST_URL = new URL("https://api.github.com/gists");
```
Comment (simon816) [inline src/main/java/org/spongepowered/common/service/error/GistPostingCallable.java]: Could use `HttpAuthenticationService.constantURL` here

File: src/main/java/org/spongepowered/common/mixin/core/server/MixinDedicatedServer.java [hunk 4]
```diff
@@ -66,6 +66,7 @@ public void setGuiEnabled() {
         this.guiIsEnabled = false;
     }
 
+
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/server/MixinDedicatedServer.java]: Remove this extra line

**Comment by gabizou** [comment]
Please update the target branch to bleeding (Edit this PR and change the target branch) and reopen.

**Pull request closed without merging**